### ISSUE TITLE
required fields

### ISSUE THREAD

**Issue by mikila85** [ajv-validator/ajv#661]
Hi,
We are using AJV to yield error messages from our user filled forms in conjunction with JSON Schema that we receive from our servers, it is working great.

One thing we weren't able to solve:
Is there a way to get all the required fields (even if filled)?

we need to mark them with * as "required" but we can't rely on the required error as it disappears after a value has the data has been filled, also the required fields might change depending on the schema and data combinations.

**Comment by epoberezkin** [comment]
Sorry, I don’t understand the question. Do you mean to get the list of fields required by the schema?

**Comment by mikila85** [comment]
yes, but only after compiling the schema with the data (because the data might affect which fields are required)

**Comment by epoberezkin** [comment]
Ajv compiles schema without the data, the data is just passed to a compiled validation function.
In general case it is not a trivial problem, as you would expect only "required" in valid schema branches taken into account.
I believe what you want is ["annotation collection" process from draft-07 spec](http://json-schema.org/latest/json-schema-validation.html#rfc.section.3.3) applied to "required" keyword.
Ajv doesn't implement collection of annotations.
Also, "required" is not an annotation keyword, although I don't see why it can't be - you can submit this proposal to [JSON Schema org](https://github.com/json-schema-org/json-schema-spec).

**Comment by mikila85** [comment]
I tried adding to the code where it locates "required" and then tries to check if there is value to addError, just before it adding similar to ajv.errors to have ajv.required.
having an array of dataPath+Property that is known to be required.

unfortunately, I have never worked with Javascript Templates and didn't succeed.
if you could help me with what to add where for this simple functionality it would solve us the problem at hand and be really grateful.

**Comment by epoberezkin** [comment]
That would simply collect all required fields in the schema, disregarding whether a given subschema passed validation.
If that's what you want it can be done once by simply traversing the schema (e.g. with json-schema-traverse).
What I believe you need would require a more complex mechanism to first take note and later to  exclude "required" properties from subschemas that failed validation because of some other reason.

**Comment by mikila85** [comment]
Exactly! schema+data should output required fields regardless of them being filled or not.
(because data can result in a field being required or not like an option box above).

Is this something you think can be added in a few lines of code (even dirty) just for it to work.
without it, the form validation feature is not complete and the release is dead in the water.

**Comment by epoberezkin** [comment]
I would do it all outside of Ajv.
1. pre-process the schema using json-schema-traverse to collect all usages of "required" keyword together with the properties (& data paths) they apply to.
2. run validation with `allErrors: true` option.
3. get the "required" you need by excluding the properties for schema branches that fail validation (for any other reason than "required") from the list collected during pre-processing step.

The only problem here is that anyOf short-circuits and you won't know whether some branches of anyOf have errors (apart from required errors). So to allow that process there should be an option to prevent short-circuiting anyOf.
Probably it's worth doing it by default but it can only be done in the next major version.

**Comment by epoberezkin** [comment]
https://github.com/epoberezkin/ajv/issues/662

**Comment by epoberezkin** [comment]
It still wouldn't work for anyOf/oneOf - ajv doesn't tell you which anyOf branches pass when the whole keyword passes, only which branches failed...

**Comment by mikiTestim** [comment]
Complicated issue.

I'm still convinced that there's a way to extract the active required-fields from the schema+data and I will keep trying to find the way.

I will surely update if/when I solve it and how for future reference to other community members.

If you think of any idea how to solve this I would appreciate it dearly!

**Comment by epoberezkin** [comment]
> there's a way to extract the active required-fields from the schema+data

There definitely is, it just requires a very different process from the one ajv uses for validation.

**Comment by Relequestual** [comment]
There may well be another library which does this! There are libraries which generate forms from JSON Schema files!

**Comment by donaldpipowitch** [comment]
Just searched for this as well and hoped `ajv` would offer an utility function like this. AFAIK there is no other big lib out there with draft-07 support and I'd like to know for a schema which uses `if/else`, if a certain field is required (given the schema and current data). (Something like `const result = isRequired(schema, data, 'some.nested.prop'); // === true or false`)

**Comment by epoberezkin** [comment]
This question cannot be answered in general case even for a defined schema and data (unless you only consider a subset of all possible JSON schemas).

**Comment by bennieswart** [comment]
We have very similar requirements. Let me just restate our use case for clarity:

We have an html form to be filled out. The underlying data is validated using a json schema with ajv. We want to be able to dynamically, based on the current json schema and data, visually mark certain fields as required with an asterisk.
This is more complicated than simply scanning for `required` keywords, as certain json schema branches may become active/inactive based on the actual data. Whatever the current subset of the json schema being applied, we want to have a list of all the required fields, regardless of whether they empty/filled in, or valid/invalid.

I created a patch some time ago which does this, which is available here: https://github.com/bennieswart/ajv/commit/4e73216e32840766633682bf7b28eed82bf623ee. The fork is outdated, but we have had no reason to upgrade yet. The same code might still work on the latest branch, you will have to check.

When all is said and done, it basically allows you to access all the required fields through `ajv.allrequired`, just as you would access all the errors through `ajv.errors`.

@epoberezkin if you think this is useful, give some feedback and I'll make a PR.

**Comment by mikila85** [comment]
Hi@bennieswart !
Thank you for taking the time to share this with us.
We have the exact needs that you have with form validation and the asterisk* for required fields and we needed allRequired too but AJV didnt supply.
We achieved it by running the AJV again and again each time emptying 1 field and seeing if it throws error of "required field". 
This is a lame solution but it works.

I wish AJV will merge the pull request from your fork,
Meanwhile we will try to work straight on your fork.
from what i understand it is not updated but your changes are limited to a small function and i think it wouldent be a problem to update.
Can you let me know if you are going to update it?

@epoberezkin, Are you willing to consider merging this valuable addition to the AJV?

**Comment by bennieswart** [comment]
@mikila85 due to your request I decided to rebase the patch onto the latest master and make a preliminary PR here: https://github.com/epoberezkin/ajv/pull/933. If you want to see this functionality added then you can go cheer it on.
In the meantime, you should be able to build from that PR, although you might want to wait until changes are finalized.
